code/controllers/subsystem/processing/greyscale.dm

```diff
@@ -107,4 +116,66 @@ PROCESSING_SUBSYSTEM_DEF(greyscale)
 	for(var/color in 2 to length(split_colors))
 		. += "#[split_colors[color]]"
 
+/datum/controller/subsystem/processing/greyscale/proc/ExportMapPreviews()
+	// Put subtypes before their parent or the parent file will take all the generated icons
+	var/static/list/types_that_get_their_own_file = list(
+		"turfs" = /turf, // None of these yet but it's harmless to be prepared
+		"mobs" = /mob, // Ditto
+		"clothing/accessory" = /obj/item/clothing/accessory,
+		"clothing/head" = /obj/item/clothing/head,
+		"clothing/mask" = /obj/item/clothing/mask,
+		"clothing/neck" = /obj/item/clothing/neck,
+		"clothing/shoes" = /obj/item/clothing/shoes,
+		"clothing/suit" = /obj/item/clothing/suit,
+		"clothing/under" = /obj/item/clothing/under,
+		"clothing/clothing" = /obj/item/clothing,
+		"items/encryptionkey" = /obj/item/encryptionkey,
+		"items/pda" = /obj/item/modular_computer/pda,
+		"items/item" = /obj/item,
+		"objects" = /obj,
+)
+
+	var/list/handled_types = list()
+	for(var/filename in types_that_get_their_own_file)
+		var/type_to_export = types_that_get_their_own_file[filename]
+		handled_types += ExportMapPreviewsForType(filename, type_to_export, handled_types)
+
+	ExportMapPreviewsForType("unsorted", /atom, handled_types)
+
+/datum/controller/subsystem/processing/greyscale/proc/ExportMapPreviewsForType(filename, atom/atom_typepath, list/type_blacklist)
+	var/list/handled_types = list()
+	var/list/icons = list()
+	for(var/atom/fake as anything in subtypesof(atom_typepath))
+		if(type_blacklist && type_blacklist[fake])
+			continue
+		handled_types[fake] = TRUE
+		var/greyscale_config = fake::greyscale_config
+		var/greyscale_colors = fake::greyscale_colors
+		if(!greyscale_config || !greyscale_colors || fake::does_not_generate_gags_preview)
+			continue
+		var/icon/map_icon = icon(GetColoredIconByType(greyscale_config, greyscale_colors))
+		if((map_icon.Height() > 32) || (map_icon.Width() > 32)) // No large icons, use icon_preview and icon_preview_state instead.
+			continue
+		if(!(fake::post_init_icon_state in map_icon.IconStates()))
+			stack_trace("GAGS configuration missing icon state needed to generate mapping tool graphic for '[fake]'. Make sure the right greyscale_config is set up.")
+			continue
+		map_icon = icon(map_icon, fake::post_init_icon_state)
+		icons["[fake]"] = map_icon
+
+	var/icon/holder = icon('icons/testing/greyscale_error.dmi')
+	for(var/state in icons)
+		holder.Insert(icons[state], state)
```

Comment:
it would be **much** faster to use `rustg_iconforge_generate_spritesheet` here. It actually natively supports GAGS icons (use `gags_to_universal_icon` to convert GAGS items to `/datum/universal_icon`s). Building large sheets of icons is literally its *entire* purpose. It can also run asynchronously on another thread and not block the master controller while it executes. It's also multithreaded. `Insert` is a very slow proc.

Reply:
Ah. One issue, that outputs PNGs, not DMIs. Nevertheless... trivial rust change, if you want me to PR some logic to make it spit out a DMI instead. It might delay this PR a bit though.